scvi/inference/posterior_utils.py

```diff
@@ -0,0 +1,291 @@
+    plt.ylabel("Imputed counts")
+    plt.xlabel("Original counts")
+
+    plt.pcolormesh(yi, xi, zi.reshape(xi.shape), cmap="Reds")
+
+    a, _, _, _ = np.linalg.lstsq(y[:, np.newaxis], x, rcond=-1)
+    linspace = np.linspace(0, ymax)
+    plt.plot(linspace, a * linspace, color="black")
+
+    plt.plot(linspace, linspace, color="black", linestyle=":")
+    if show_plot:
+        plt.show()
+    plt.savefig(title + ".png")
+
+
+def nn_overlap(X1, X2, k=100):
+    """
+    Compute the overlap between the k-nearest neighbor graph of X1 and X2 using Spearman correlation of the
+    adjacency matrices.
+    """
+    assert len(X1) == len(X2)
+    n_samples = len(X1)
+    k = min(k, n_samples - 1)
+    nne = NearestNeighbors(n_neighbors=k + 1)  # "n_jobs=8
+    nne.fit(X1)
+    kmatrix_1 = nne.kneighbors_graph(X1) - scipy.sparse.identity(n_samples)
+    nne.fit(X2)
+    kmatrix_2 = nne.kneighbors_graph(X2) - scipy.sparse.identity(n_samples)
+
+    # 1 - spearman correlation from knn graphs
+    spearman_correlation = scipy.stats.spearmanr(
+        kmatrix_1.A.flatten(), kmatrix_2.A.flatten()
+    )[0]
+    # 2 - fold enrichment
+    set_1 = set(np.where(kmatrix_1.A.flatten() == 1)[0])
+    set_2 = set(np.where(kmatrix_2.A.flatten() == 1)[0])
+    fold_enrichment = (
+        len(set_1.intersection(set_2))
+        * n_samples ** 2
+        / (float(len(set_1)) * len(set_2))
+    )
+    return spearman_correlation, fold_enrichment
+
+
+def unsupervised_clustering_accuracy(y, y_pred):
+    """
+    Unsupervised Clustering Accuracy
+    """
+    assert len(y_pred) == len(y)
+    u = np.unique(np.concatenate((y, y_pred)))
+    n_clusters = len(u)
+    mapping = dict(zip(u, range(n_clusters)))
+    reward_matrix = np.zeros((n_clusters, n_clusters), dtype=np.int64)
+    for y_pred_, y_ in zip(y_pred, y):
+        if y_ in mapping:
+            reward_matrix[mapping[y_pred_], mapping[y_]] += 1
+    cost_matrix = reward_matrix.max() - reward_matrix
+    ind = linear_assignment(cost_matrix)
+    return sum([reward_matrix[i, j] for i, j in ind]) * 1.0 / y_pred.size, ind
+
+
+def knn_purity(latent, label, n_neighbors=30):
+    nbrs = NearestNeighbors(n_neighbors=n_neighbors + 1).fit(latent)
+    indices = nbrs.kneighbors(latent, return_distance=False)[:, 1:]
+    neighbors_labels = np.vectorize(lambda i: label[i])(indices)
+
+    # pre cell purity scores
+    scores = ((neighbors_labels - label.reshape(-1, 1)) == 0).mean(axis=1)
+    res = [
+        np.mean(scores[label == i]) for i in np.unique(label)
+    ]  # per cell-type purity
+
+    return np.mean(res)
+
+
+def proximity_imputation(real_latent1, normed_gene_exp_1, real_latent2, k=4):
+    knn = KNeighborsRegressor(k, weights="distance")
+    y = knn.fit(real_latent1, normed_gene_exp_1).predict(real_latent2)
+    return y
+
+
+def pairs_sampler(
+    arr1: Union[List[float], np.ndarray, torch.Tensor],
+    arr2: Union[List[float], np.ndarray, torch.Tensor],
+    sample_pairs: bool = True,
```

Comment:
It's confusing that this function is called `pairs_sampler` and that there's a `sample_pairs` option. Is it possible to clarify this? Like, `sample_pairs` activates the permutation, so why not call it `use_permutation`?